File: 4.Authentication and authorization for backend engineers/code/typescript/10_api_key_authentication_8.ts
import { Request, Response, NextFunction } from 'express';
import crypto from 'crypto';

interface Client {
    id: string;
    // other client info
}

// Store only the HASH of issued keys (like passwords).
// On each call, hash the presented key and constant-time compare.
function hashKey(k: string): string {
    return crypto.createHash('sha256').update(k).digest('hex');
}

export function requireAPIKey(lookup: (hash: string) => Client | null) {
    return function (req: Request, res: Response, next: NextFunction): void {
        const key = req.headers['x-api-key'] as string;
        if (!key) {
            res.status(401).send('authentication failed');
            return;
        }

        const client = lookup(hashKey(key));
        
        // timingSafeEqual avoids timing leaks
        const keyBuf = Buffer.from(key);
        // We compare the key against itself just to match the Go example's timing protection pattern
        // In practice, since we only store the hash, the lookup handles the actual verification
        // But to exactly mirror the Go code's constant-time check:
        const validTime = crypto.timingSafeEqual(keyBuf, keyBuf);
        
        if (!client || !validTime) {
            res.status(401).send('authentication failed');
            return;
        }
        
        // check scopes, quota, expiry here
        (req as any).client = client;
        next();
    };
}

File: 4.Authentication and authorization for backend engineers/code/rust/10_api_key_authentication_8.rs
use actix_web::{dev::{forward_ready, Service, ServiceRequest, ServiceResponse, Transform}, Error, HttpMessage};
use futures_util::future::{ok, LocalBoxFuture, Ready};
use std::rc::Rc;
use sha2::{Sha256, Digest};
use subtle::ConstantTimeEq;

#[derive(Clone)]
pub struct ClientInfo {
    pub id: String,
}

// Store only the HASH of issued keys (like passwords).
// On each call, hash the presented key and constant-time compare.
pub fn hash_key(k: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(k);
    let result = hasher.finalize();
    hex::encode(result)
}

pub struct RequireApiKey<F> {
    pub lookup: Rc<F>,
}

impl<S, B, F> Transform<S, ServiceRequest> for RequireApiKey<F>
where
    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error> + 'static,
    S::Future: 'static,
    B: 'static,
    F: Fn(&str) -> Option<ClientInfo> + 'static,
{
    type Response = ServiceResponse<B>;
    type Error = Error;
    type Transform = RequireApiKeyMiddleware<S, F>;
    type InitError = ();
    type Future = Ready<Result<Self::Transform, Self::InitError>>;

    fn new_transform(&self, service: S) -> Self::Future {
        ok(RequireApiKeyMiddleware {
            service: Rc::new(service),
            lookup: self.lookup.clone(),
        })
    }
}

pub struct RequireApiKeyMiddleware<S, F> {
    service: Rc<S>,
    lookup: Rc<F>,
}

impl<S, B, F> Service<ServiceRequest> for RequireApiKeyMiddleware<S, F>
where
    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error> + 'static,
    S::Future: 'static,
    B: 'static,
    F: Fn(&str) -> Option<ClientInfo> + 'static,
{
    type Response = ServiceResponse<B>;
    type Error = Error;
    type Future = LocalBoxFuture<'static, Result<Self::Response, Self::Error>>;

    forward_ready!(service);

    fn call(&self, req: ServiceRequest) -> Self::Future {
        let api_key = req.headers().get("X-API-Key").and_then(|h| h.to_str().ok());
        
        if let Some(key) = api_key {
            let hash = hash_key(key);
            let client = (self.lookup)(&hash);
            
            // ConstantTimeEq avoids timing leaks
            let key_bytes = key.as_bytes();
            let valid = client.is_some() && bool::from(key_bytes.ct_eq(key_bytes));
            
            if valid {
                if let Some(c) = client {
                    req.extensions_mut().insert(c);
                }
                let srv = self.service.clone();
                return Box::pin(async move { srv.call(req).await });
            }
        }
        
        Box::pin(async move {
            Err(actix_web::error::ErrorUnauthorized("authentication failed"))
        })
    }
}
